File: core/MostCommonWord/MostCommonWord.java
package com.dbc;

import java.util.*;

public class MostCommonWord {
    public String mostCommonWord(String paragraph, String[] banned) {
        Set<String> hit = new HashSet<>();
        Map<String, Integer> counter = new HashMap<>();
        for (String temp : banned) hit.add(temp);
        int left = 0, right = 0;
        String res = "";
        while (right < paragraph.length()) {
            while (right < paragraph.length() && Character.isLetter(paragraph.charAt(right))) right++;
            String letter = paragraph.substring(left, right).toLowerCase();
            counter.put(letter, counter.getOrDefault(letter, 0) + 1);
            if (counter.get(letter) > counter.getOrDefault(res, 0) && !hit.contains(letter)) res = letter;
            while (right < paragraph.length() && !Character.isLetter(paragraph.charAt(right))) right++;
            left = right;
        }
        return res;
    }
}

File: core/MostCommonWord/MostCommonWord.py
from typing import List
from collections import Counter

def mostCommonWord(paragraph: str, banned: List[str]) -> str:
    p_char = ["!", "?", "'", ",", ";", "."]
    left, right, length = 0, 0, len(paragraph)
    res, counter = "", Counter()

    while right < length:
        while right < length and paragraph[right].isalpha():
            right += 1
        ch = paragraph[left:right].lower()
        counter[ch] += 1
        if counter[ch] > counter[res] and ch not in banned:
            res = ch

        while right < length and not paragraph[right].isalpha():
            right += 1
        left = right

    return res


if __name__ == '__main__':
    print(mostCommonWord("Bob hit a ball, the hit BALL flew far after it was hit.", ["hit"]))







